public/app/features/alerting/unified/components/rules/central-state-history/utils.ts

```diff
@@ -0,0 +1,138 @@
+import { groupBy } from 'lodash';
+
+import { DataFrame, Field as DataFrameField, DataFrameJSON, Field, FieldType } from '@grafana/data';
+import { fieldIndexComparer } from '@grafana/data/src/field/fieldComparers';
+
+import { labelsMatchMatchers, parseMatchers } from '../../../utils/alertmanager';
+import { LogRecord } from '../state-history/common';
+import { isLine, isNumbers } from '../state-history/useRuleHistoryRecords';
+
+import { LABELS_FILTER } from './CentralAlertHistoryScene';
+
+const GROUPING_INTERVAL = 10 * 1000; // 10 seconds
+const QUERY_PARAM_PREFIX = 'var-'; // Prefix used by Grafana to sync variables in the URL
+/*
+ * This function is used to convert the history response to a DataFrame list and filter the data by labels.
+ * The response is a list of log records, each log record has a timestamp and a line.
+ * We group all records by alert instance (unique set of labels) and create a DataFrame for each group (instance).
+ * This allows us to be able to filter by labels in the groupDataFramesByTime function.
+ */
+export function historyResultToDataFrame(data: DataFrameJSON): DataFrame[] {
+  const tsValues = data?.data?.values[0] ?? [];
+  const timestamps: number[] = isNumbers(tsValues) ? tsValues : [];
+  const lines = data?.data?.values[1] ?? [];
+
+  const logRecords = timestamps.reduce((acc: LogRecord[], timestamp: number, index: number) => {
+    const line = lines[index];
+    // values property can be undefined for some instance states (e.g. NoData)
+    if (isLine(line)) {
+      acc.push({ timestamp, line });
+    }
+    return acc;
+  }, []);
+
+  // Group log records by alert instance
+  const logRecordsByInstance = groupBy(logRecords, (record: LogRecord) => {
+    return JSON.stringify(record.line.labels);
+  });
+
+  // Convert each group of log records to a DataFrame
+  const dataFrames: DataFrame[] = Object.entries(logRecordsByInstance).map<DataFrame>(([key, records]) => {
+    // key is the stringified labels
+    return logRecordsToDataFrame(key, records);
+  });
+
+  // Group DataFrames by time and filter by labels
+  return groupDataFramesByTimeAndFilterByLabels(dataFrames);
+}
+
+// Scenes sync variables in the URL adding a prefix to the variable name.
+function getFilterInQueryParams() {
+  const queryParams = new URLSearchParams(window.location.search);
```

Comment:
Could we use `useLocation`? I think that's what we're doing elsewhere

Reply:
We cannot, as this function is called from a class component (no hooks allowed)